### PR TITLE
ReduceScatter for compute p2p communication pattern

### PR DESCRIPTION
**Pull request by eldarkh**
This is an attempt to use `MPI_Reduce_scatter_block` for the implementation of `Utilities::MPI::compute_point_to_point_communication_pattern` as discussed in [#6796](https://github.com/dealii/dealii/issues/6796)

The unit test is unchanged, apart from sorting of the resulting vector for reproducibility, as now there is nothing that orders it implicitly as with previous implementation.

### CODE REVIEW

File: source/base/mpi.cc [hunk 1]
```diff
@@ -197,50 +198,42 @@ namespace Utilities
...
-            origins.push_back(i);
-          else if (all_destinations[i * max_n_destinations + j] ==
-                   numbers::invalid_unsigned_int)
-            break;
+      // If no one to receive from, return an empty vector
+      if (n_recv_from == 0)
+        return std::vector<unsigned int>();
 
+      // Send myid to every process in `destinations` vector...
+      for (auto &el : destinations)
```
Comment (masterleinad) [inline source/base/mpi.cc]: Can you make `el` a constant reference? In the end, you are not modifying it.

File: source/base/mpi.cc [hunk 2]
```diff
@@ -197,50 +198,42 @@ namespace Utilities
...
+        MPI_Send(&myid, 1, MPI_UNSIGNED, el, myid, mpi_comm);
+
+      // ...and receive `n_recv_from` times from the processes
+      // who communicate with this one. Store the obtained id's
+      // in the resulting vector
+      std::vector<unsigned int> origins(n_recv_from);
+      for (unsigned int i = 0; i < n_recv_from; ++i)
```
Comment (masterleinad) [inline source/base/mpi.cc]: You could use a range-based loop here as well.

File: source/base/mpi.cc [hunk 3]
```diff
@@ -197,50 +198,42 @@ namespace Utilities
...
-          else if (all_destinations[i * max_n_destinations + j] ==
-                   numbers::invalid_unsigned_int)
-            break;
+      // If no one to receive from, return an empty vector
+      if (n_recv_from == 0)
+        return std::vector<unsigned int>();
 
+      // Send myid to every process in `destinations` vector...
+      for (auto &el : destinations)
+        MPI_Send(&myid, 1, MPI_UNSIGNED, el, myid, mpi_comm);
```
Comment (masterleinad) [inline source/base/mpi.cc]: In general, calls to `MPI_Send` are blocking. You probably want to use the non-blocking equivalent `MPI_Isend` here to avoid a deadlock.

File: source/base/mpi.cc [hunk 4]
```diff
@@ -197,50 +198,42 @@ namespace Utilities
...
-      for (unsigned int i = 0; i < n_procs; ++i)
-        for (unsigned int j = 0; j < max_n_destinations; ++j)
-          if (all_destinations[i * max_n_destinations + j] == myid)
-            origins.push_back(i);
-          else if (all_destinations[i * max_n_destinations + j] ==
-                   numbers::invalid_unsigned_int)
-            break;
+      // If no one to receive from, return an empty vector
+      if (n_recv_from == 0)
+        return std::vector<unsigned int>();
```
Comment (masterleinad) [inline source/base/mpi.cc]: Would you mind to put in a `MPI_Barrier` here as well before returning?

File: source/base/mpi.cc [hunk 5]
```diff
@@ -197,50 +198,42 @@ namespace Utilities
...
+      // in the resulting vector
+      std::vector<unsigned int> origins(n_recv_from);
+      for (unsigned int i = 0; i < n_recv_from; ++i)
+        MPI_Recv(&origins[i],
+                 1,
+                 MPI_UNSIGNED,
+                 MPI_ANY_SOURCE,
+                 MPI_ANY_TAG,
+                 mpi_comm,
+                 &status);
```
Comment (masterleinad) [inline source/base/mpi.cc]: You can use `MPI_STATUS_IGNORE` here.

File: source/base/mpi.cc [hunk 6]
```diff
@@ -187,6 +187,7 @@ namespace Utilities
     {
       const unsigned int myid    = Utilities::MPI::this_mpi_process(mpi_comm);
       const unsigned int n_procs = Utilities::MPI::n_mpi_processes(mpi_comm);
+      MPI_Status         status;
```
Comment (masterleinad) [inline source/base/mpi.cc]: If you are using `MPI_STATUS_IGNORE` below instead, you don't need this variable.

File: source/base/mpi.cc [hunk 7]
```diff
@@ -187,6 +187,7 @@ namespace Utilities
...
 
       for (unsigned int i = 0; i < destinations.size(); ++i)
```
Comment (masterleinad) [inline source/base/mpi.cc]: If you feel like it, you could turn this loop into range-based one as well.

**Comment by eldarkh** [comment]
@masterleinad I just tested the performance of this implementation using the same setting as in the unit test and it is slower than the original one by a factor of 2-3 (I guess I should have started with this before opening a pull request).

On the other hand, and correct me if I am wrong here, I have noticed that when this function is used in deal.II it is only to find the **number** of processes to receive from, and this is what `MPI_Reduce_scatter` does in my implementation, and this (without consequent sends/receives) is indeed noticeably faster than what was originally implemented (see the attached table, in the last row times are larger due to communication outside of the node)
table

**Comment by masterleinad** [comment]
How exactly does the code look like you used for testing?
With the one below, the new implementation takes around 10% more time using 80 processes (this is the maximum I can use with out oversubscribing; all of this on one node). With oversubscribing the new implementation is way faster for me (factor 3 for 128 processes).
`MPI_Reduce_scatter` takes around half of the time for me.

Can you repeat your timing experiments with the code below?

```
// ---------------------------------------------------------------------
//
// Copyright (C) 2009 - 2018 by the deal.II authors
//
// This file is part of the deal.II library.
//
// The deal.II library is free software; you can use it, redistribute
// it, and/or modify it under the terms of the GNU Lesser General
// Public License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
// The full text of the license can be found in the file LICENSE.md at
// the top level directory of deal.II.
//
// ---------------------------------------------------------------------

// Check if point-to-point communication
// works correctly with MPI_Reduce_scatter_block
// implementation

#include <algorithm>
#include <iostream>
#include <mpi.h>
#include <vector>

std::vector<unsigned int> compute_point_to_point_communication_pattern_old(
    const MPI_Comm &mpi_comm, const std::vector<unsigned int> &destinations) {
  // Get the number of processes
  int n_procs;
  MPI_Comm_size(mpi_comm, &n_procs);

  // Get the rank of the process
  int myid;
  MPI_Comm_rank(mpi_comm, &myid);

  // let all processors communicate the maximal number of destinations
  // they have
  const unsigned int n_local_destinations = destinations.size();
  unsigned int max_n_destinations;
  MPI_Allreduce(&n_local_destinations, &max_n_destinations, 1, MPI_UNSIGNED,
                MPI_MAX, mpi_comm);

  if (max_n_destinations == 0)
    // all processes have nothing to send/receive:
    return std::vector<unsigned int>();

  // now that we know the number of data packets every processor wants to
  // send, set up a buffer with the maximal size and copy our destinations
  // in there, padded with -1's
  std::vector<unsigned int> my_destinations(max_n_destinations, -1);
  std::copy(destinations.begin(), destinations.end(), my_destinations.begin());

  // now exchange these (we could communicate less data if we used
  // MPI_Allgatherv, but we'd have to communicate my_n_destinations to all
  // processors in this case, which is more expensive than the reduction
  // operation above in MPI_Allreduce)
  std::vector<unsigned int> all_destinations(max_n_destinations * n_procs);
  MPI_Allgather(my_destinations.data(), max_n_destinations, MPI_UNSIGNED,
                all_destinations.data(), max_n_destinations, MPI_UNSIGNED,
                mpi_comm);

  // now we know who is going to communicate with whom. collect who is
  // going to communicate with us!
  std::vector<unsigned int> origins;
  for (int i = 0; i < n_procs; ++i)
    for (unsigned int j = 0; j < max_n_destinations; ++j)
      if (all_destinations[i * max_n_destinations + j] == (unsigned int)myid)
        origins.push_back(i);
      else if (all_destinations[i * max_n_destinations + j] == (unsigned int)-1)
        break;

  return origins;
}

std::vector<unsigned int> compute_point_to_point_communication_pattern_new(
    const MPI_Comm &mpi_comm, const std::vector<unsigned int> &destinations) {
  // Get the number of processes
  int n_procs;
  MPI_Comm_size(mpi_comm, &n_procs);

  // Get the rank of the process
  int myid;
  MPI_Comm_rank(mpi_comm, &myid);

  MPI_Status status;

  // Get a binary vector from the `destinations`
  std::vector<unsigned int> dest_vector(n_procs);
  for (auto &el : destinations)
    dest_vector[el] = 1;

  // Find how many processes will send to this one
  // by reducing with sum and then scattering the
  // results over all processes
  unsigned int n_recv_from;
  MPI_Reduce_scatter_block(&dest_vector[0], &n_recv_from, 1, MPI_UNSIGNED,
                           MPI_SUM, mpi_comm);

  // If no one to receive from, return an empty vector
  if (n_recv_from == 0)
    return std::vector<unsigned int>();

  // Send myid to every process in `destinations` vector...
  std::vector<MPI_Request> send_requests(destinations.size());
  std::vector<MPI_Status> send_stati(destinations.size());
  for (unsigned int i = 0; i < destinations.size(); ++i)
    MPI_Isend(&myid, 1, MPI_UNSIGNED, destinations[i], myid, mpi_comm,
              &send_requests[i]);

  // ...and receive `n_recv_from` times from the processes
  // who communicate with this one. Store the obtained id's
  // in the resulting vector
  std::vector<unsigned int> origins(n_recv_from);
  for (unsigned int i = 0; i < n_recv_from; ++i)
    MPI_Recv(&origins[i], 1, MPI_UNSIGNED, MPI_ANY_SOURCE, MPI_ANY_TAG,
             mpi_comm, &status);

  MPI_Waitall(destinations.size(), send_requests.data(), send_stati.data());
  //  MPI_Barrier(mpi_comm);
  return origins;
}

void test_mpi() {
  const MPI_Comm mpi_comm = MPI_COMM_WORLD;
  // Get the number of processes
  int numprocs;
  MPI_Comm_size(mpi_comm, &numprocs);

  // Get the rank of the process
  int myid;
  MPI_Comm_rank(mpi_comm, &myid);

  // select a few destinations
  std::vector<unsigned int> destinations;
  for (int i = 0; i < 3 + myid / 3; ++i)
    if ((myid + 17 * i) % numprocs != myid)
      destinations.push_back((myid + 17 * i) % numprocs);

  /*  if (myid == 0) {
      std::cout << "Processor 0 wants to send to ";
      for (unsigned int i = 0; i < destinations.size(); ++i)
        std::cout << destinations[i] << ' ';
      std::cout << std::endl;

      for (int p = 1; p < numprocs; ++p) {
        MPI_Status status;
        unsigned int size = 0;
        MPI_Recv(&size, 1, MPI_UNSIGNED, p, 0, MPI_COMM_WORLD, &status);

        std::vector<unsigned int> dest(size);
        MPI_Recv(&dest[0], size, MPI_UNSIGNED, p, 0, MPI_COMM_WORLD, &status);

        std::cout << "Processor " << p << " wants to send to ";
        for (unsigned int i = 0; i < size; ++i)
          std::cout << dest[i] << ' ';
        std::cout << std::endl;
      }
    } else {
      unsigned int size = destinations.size();
      MPI_Send(&size, 1, MPI_UNSIGNED, 0, 0, MPI_COMM_WORLD);
      MPI_Send(&destinations[0], size, MPI_UNSIGNED, 0, 0, MPI_COMM_WORLD);
    }*/

  // std::cout << "Exchanging data..." << std::endl;
  std::vector<unsigned int> origins;
  for (unsigned int i = 0; i < 1000000; ++i) {
    origins = compute_point_to_point_communication_pattern_new(MPI_COMM_WORLD,
                                                               destinations);
    if (myid == 0) {
      std::cout << origins.size() << std::endl;
    }
  }

  //  return origins.size();
  /*  std::sort(origins.begin(), origins.end());

    if (myid == 0)
    {
      std::cout << "Processor 0 will receive from ";
      for (unsigned int i = 0; i < origins.size(); ++i)
        std::cout << origins[i] << ' ';
      std::cout << std::endl;

      for (int p = 1; p < numprocs; ++p) {
        MPI_Status status;
        unsigned int size = 0;
        MPI_Recv(&size, 1, MPI_UNSIGNED, p, 0, MPI_COMM_WORLD, &status);

        std::vector<unsigned int> orig(size);
        MPI_Recv(&orig[0], size, MPI_UNSIGNED, p, 0, MPI_COMM_WORLD, &status);

        std::cout << "Processor " << p << " will receive from ";
        for (unsigned int i = 0; i < size; ++i)
          std::cout << orig[i] << ' ';
        std::cout << std::endl;
      }
    } else {
      unsigned int size = origins.size();
      MPI_Send(&size, 1, MPI_UNSIGNED, 0, 0, MPI_COMM_WORLD);
      MPI_Send(&origins[0], size, MPI_UNSIGNED, 0, 0, MPI_COMM_WORLD);
    }*/
}

int main(int argc, char *argv[]) {
  MPI_Init(&argc, &argv);

  test_mpi();

  // std::cout << n_return << std::endl;

  MPI_Finalize();
}
```

**Comment by eldarkh** [comment]
@masterleinad Using your code to test the timings I get the results that agree with what you got. My test was very similar, although I used `MPI_Barrier` instead of `MPI_Waitall`, so this might explain the difference in performance... Also my initial logic with early return was slightly wrong, so I moved it below the `MPI_Isend` block, since having no one to receive from does not mean that there is no one to send to. Even, with this change made to your timing test, the results agree with what you got.

I will incorporate the changes that you requested earlier and commit them shortly. Thanks for your feedback.

File: source/base/mpi.cc [hunk 8]
```diff
@@ -197,7 +197,54 @@ namespace Utilities
...
+      // ...otherwise receive `n_recv_from` times from the processes
+      // who communicate with this one. Store the obtained id's
+      // in the resulting vector
+      std::vector<unsigned int> origins(n_recv_from);
+      for (auto &el : origins)
+        MPI_Recv(&el,
+                 1,
+                 MPI_UNSIGNED,
+                 MPI_ANY_SOURCE,
+                 1 << 26,
```
Comment (masterleinad) [inline source/base/mpi.cc]: See above.

File: source/base/mpi.cc [hunk 9]
```diff
@@ -197,7 +197,54 @@ namespace Utilities
...
+      AssertThrowMPI(ierr);
+
+      // Send myid to every process in `destinations` vector...
+      std::vector<MPI_Request> send_requests(destinations.size());
+      for (const auto &el : destinations)
+        MPI_Isend(&myid,
+                  1,
+                  MPI_UNSIGNED,
+                  el,
+                  1 << 26,
```
Comment (masterleinad) [inline source/base/mpi.cc]: This value is too large in general, the MPI standard requires MPI_TAG_UB (the maximum allowed tag) to be at least 32767 (2^15-1). Just pick something below that.

**Comment by jppelteret** [comment]
@masterleinad Should this be merged?

**Comment by masterleinad** [comment]
I am not quite sure what the performance impact is and if it is relevant at all in a typical code.
Hence, I didn't want to decide this alone.

**Comment by jppelteret** [comment]
Fair enough :-) I'm not sufficiently proficient with MPI to comment on this, but I hope that someone else would be able to give a second opinion soon.

File: source/base/mpi.cc [hunk 10]
```diff
@@ -658,4 +706,4 @@ namespace Utilities
   } // end of namespace MPI
 } // end of namespace Utilities
 
-DEAL_II_NAMESPACE_CLOSE
+DEAL_II_NAMESPACE_CLOSE
```
Comment (kronbichler) [inline source/base/mpi.cc]: Can you please undo this change (no newline at end of line)?

**Comment by eldarkh** [comment]
@kronbichler I made the change as you suggested

**Comment by kronbichler** [comment]
This is ready to merge now.

**Comment by bangerth** [comment]
@eldarkh -- coming late to the game here, but thanks for tackling this!

You mentioned that all of the places where the current function is called actually only need to know how many, not which senders there will be. Do you think there is a point in providing a separate function that computes only that, and that the current callers can call?

**Comment by eldarkh** [comment]
@bangerth I think this might be worth it with newer MPI, since the speed up will be more significant than what I did now, and it is quite easy to implement. I am not yet sure how the old implementation, can be modified to handle this simpler task only, I will try mimicking what's done in `Reduce_scatter` and see if it is worth it performance-wise.

I also wanted to ask, which MPI documentation should I use as reference? The reason I ask is it seems that OpenMPI and MPICH has `Reduce_scatter_block` in versions prior to 2.0 even, on the other hand according to the documentation on [mpi-forums.org](https://www.mpi-forum.org/docs/) `Reduce_scatter` is available in versions prior to 2.0, but not `Reduce_scatter_block`, and I don't know about Intel.. So I just want to make sure which tools I can and cannot use.

**Comment by bangerth** [comment]
@eldarkh: I think the canonical documentation is on mpi-forum.org.

It would of course be excellent if you want to implement a separate function and see whether it can be used in a couple of places. Feel free to first open a github issue that tracks what you want to do and that can be closed once a pull request is merged!

**Comment by masterleinad** [comment]
I created #6954 so we don't forget about implementing such a function.
